### PR TITLE
Merge from hdinsight to Dev, and apply several bug fixes on hdinsight._j...

### PR DESCRIPTION
**Pull request by GuoyuHao**
...s

1. Merge back to Dev, hdinsight was used to ship OOB hdinsight package on
Feb, we need to keep working on Dev in future

2. There are several bug fixes against Feb's release, apply them on Dev branch

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@GuoyuHao__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

File: lib/commands/asm/hdinsight._js [hunk 1]
```diff
@@ -534,107 +967,132 @@ hdInsightCommandLine.init = function (cli) {
   var self = new hdInsightCommandLine(cli);
 
   var hdInsight = cli.category('hdinsight')
-    .description($('Commands to manage your HDInsight accounts'));
+    .description($('PREVIEW: Commands to manage your HDInsight clusters and jobs'));
```
Comment (amarzavery) [inline lib/commands/asm/hdinsight._js]: Is this still Preview. If not can you please remove  this?

File: lib/commands/asm/hdinsight._js [hunk 2]
```diff
@@ -534,107 +967,132 @@ hdInsightCommandLine.init = function (cli) {
...
+      .option('--password <password>', $('Password for the cluster'))
+      .option('--query <query>', $('The Hive query string to be executed'))
+      .option('--queryFile <queryFile>', $('The path to a file that contains the Hive query to be executed; this parameter and the [query] parameter are mutually exclusive'))
+      .option('--arguments <arguments>', $('A comma separated string of arguments to be passed to the Hive job. For example: "a1,a2,a3"'))
+      .option('--defines <defines>', $('A key/value pair of Hadoop configuration values to be set during the Hive job execution. For example: "k1=v1,k2=v2"'))
+      .option('--files <files>', $('A comma separated string of file paths required for the Hive job to execute. For example: "f1/f2/f3,f4/f5,f6"'))
+      .option('-s, --subscription <id>', $('the subscription id'))
+      .execute(self.submitHDInsightHiveJobCommand);
+
+    /*job.command('pig_create [clusterDnsName] [userName] [password] [query]')
```
Comment (amarzavery) [inline lib/commands/asm/hdinsight._js]: Can you also enable pig_create command?

File: lib/commands/asm/hdinsight._js [hunk 3]
```diff
@@ -259,272 +246,718 @@ var hdInsightCommandLine = function(cli, userInteractor, executionProcessor) {
...
+    location = self.user.promptIfNotGiven($('Data center location: '), location, _);
+    userName = self.user.promptIfNotGiven($('User name: '), userName, _);
+    password = self.user.promptIfNotGiven($('Password: '), password, _);
+
+    var clusterCreationPayload = {};
+    if(osType === 'windows')
+    {
+      headNodeSize = self.user.promptIfNotGiven($('Head node size (string): '), headNodeSize, _);
+      dataNodeSize = self.user.promptIfNotGiven($('Data node size (string): '), dataNodeSize, _);
+      var clusterPayload = '<Resource xmlns="http://schemas.microsoft.com/windowsazure">';
```
Comment (amarzavery) [inline lib/commands/asm/hdinsight._js]: can this gigantic string be put into a separate file and reference from there.

Comment (GuoyuHao) [inline lib/commands/asm/hdinsight._js]: This has been updated. https://github.com/Azure/azure-xplat-cli/commit/56abf2ad6e06809d0403431a2d95c25a5e1b88a8

File: lib/commands/asm/hdinsight._js (hunk 1 above)
Comment (GuoyuHao) [inline lib/commands/asm/hdinsight._js]: This has been updated. https://github.com/Azure/azure-xplat-cli/commit/56abf2ad6e06809d0403431a2d95c25a5e1b88a8

File: lib/commands/asm/hdinsight._js (hunk 2 above)
Comment (GuoyuHao) [inline lib/commands/asm/hdinsight._js]: This part is not ready yet. We'd like to keep this commented out, and adding the support later in another pull request. 
Actually there might be more feature gaps other than pig job supporting, we will review them all together, and have a plan to deliver them.

File: test/hdinsight/stub-cli.js [hunk 4]
```diff
@@ -15,15 +15,16 @@
 // 
 
 var sinon = require('sinon');
-var Interaction = require('../../lib/util/interaction');
+//var Interaction = require('../../lib/util/interaction');
```
Comment (yugangw-msft) [inline test/hdinsight/stub-cli.js]: could you just remove it?

File: test/hdinsight/stub-cli.js [hunk 5]
```diff
@@ -15,15 +15,16 @@
...
 
 var CliStub = function() {
   this.categories = { };
   this.commands = { };
   this.options = { };
   this.optionCount = 0;
-  this.interaction = new Interaction(this);
+  //this.interaction = new Interaction(this);
```
Comment (yugangw-msft) [inline test/hdinsight/stub-cli.js]: same here

File: test/hdinsight/stub-cli.js [hunk 6]
```diff
@@ -15,15 +15,16 @@
...
 
+  this.output = sinon.spy();
```
Comment (yugangw-msft) [inline test/hdinsight/stub-cli.js]: also any reasons we want to switch from interaction to sinon?

File: test/hdinsight/stub-userInteractor.js [hunk 7]
```diff
@@ -61,7 +61,8 @@ var UserInteractor = function() {
     'Metastore Server: ' : 'server1',
     'Metastore Database: ' : 'database1',
     'Metastore user: ' : 'user1',
-    'Metastore password: ' : 'password'
+    'Metastore password: ' : 'password',
+	'Data center location: ' : 'East US'
```
Comment (yugangw-msft) [inline test/hdinsight/stub-userInteractor.js]: indent

File: test/hdinsight/unit-cli-setup.js [hunk 8]
```diff
@@ -136,11 +125,13 @@ describe('HDInsight command line (under unit test)', function() {
...
     var command = cluster.commands['show'];
-    command.optionCount.should.be.equal(2);
+    command.optionCount.should.be.equal(3);
     command.options['-s, --subscription <id>'].should.not.equal(null);
-    command.options['-s, --subscription <id>'].should.be.equal('the subscription id')
+    command.options['-s, --subscription <id>'].should.be.equal('the subscription id');
     command.options['--clusterName <clusterName>'].should.not.equal(null);
-    command.options['--clusterName <clusterName>'].should.be.equal('the HdInsight cluster name')
+    command.options['--clusterName <clusterName>'].should.be.equal('the HdInsight cluster name');
+	command.options['--osType <osType>'].should.not.equal(null);
```
Comment (yugangw-msft) [inline test/hdinsight/unit-cli-setup.js]: indent

File: lib/commands/asm/hdinsight._js [hunk 9]
```diff
@@ -259,272 +247,373 @@ var hdInsightCommandLine = function(cli, userInteractor, executionProcessor) {
...
+      parameters.file = options.queryFile;
+    }
+    if (options['arguments']) {
+      parameters['arguments'] = options['arguments'];
+    }
+    if (options.files) {
+      parameters.files = options.files;
+    }
+    var result = self.processor.submitHDInsightPigJob(clusterDnsName, userName, parameters, _);
+    console.log(util.inspect(result, { depth: null }));
```
Comment (yugangw-msft) [inline lib/commands/asm/hdinsight._js]: you might need to remove this

File: lib/commands/asm/hdinsight._js [hunk 10]
```diff
@@ -259,272 +247,373 @@ var hdInsightCommandLine = function(cli, userInteractor, executionProcessor) {
...
     }
   };
 
-  this.showClusterCommand = function (clusterName, options, _) {
+  this.showClusterCommand = function (clusterName, osType, options, _) {
     clusterName = self.user.promptIfNotGiven($('Cluster name: '), clusterName, _);
-    self.user.startProgress($('Getting HDInsight cluster'));
-    var cluster = self.processor.getCluster(clusterName, options.subscription, _);
+    osType = self.user.promptIfNotGiven($('OS type: '), osType, _);
+    //location = self.user.promptIfNotGiven($('Location: '), location, _);
```
Comment (yugangw-msft) [inline lib/commands/asm/hdinsight._js]: remove this

File: lib/commands/asm/hdinsight._js [hunk 11]
```diff
@@ -259,272 +247,373 @@ var hdInsightCommandLine = function(cli, userInteractor, executionProcessor) {
...
+    var clusterCreationPayload = {};
+    if (osType === 'windows') {
+      headNodeSize = self.user.promptIfNotGiven($('Head node size (string): '), headNodeSize, _);
+      dataNodeSize = self.user.promptIfNotGiven($('Data node size (string): '), dataNodeSize, _);
+
+      var clusterPayload = hdiHelper.createClusterPayloadWindows(clusterName, storageAccountName, storageAccountKey, storageContainer, dataNodeCount, headNodeSize, dataNodeSize, location, userName, password);
+
+      clusterCreationPayload = { 'payload': clusterPayload };
     }
+    else if (osType == 'linux') {
```
Comment (yugangw-msft) [inline lib/commands/asm/hdinsight._js]: use ===

File: test/testlist.txt [hunk 12]
```diff
@@ -81,12 +81,11 @@ commands/cli.storage.queue-tests.js
 commands/cli.storage.service-tests.js
 util/git/linkrevisioncontrol-tests.js
 # To remove once the new hdinsight commands checked in
```
Comment (yugangw-msft) [inline test/testlist.txt]: remove this comment

**Comment by yugangw-msft** [comment]
error caused by 0.8 node npm installation failure

**Pull request merged**